mlflow/transformers/__init__.py

```diff
@@ -2928,6 +2975,59 @@ def _format_prompt_template(self, input_data):
             "Prompt templating is only supported for data of type str or List[str]. "
         )
 
+    def _postprocess_output_for_inference_task(self, data, output):
+        """
+        Compute the token usage and finish reason for the output of a text generation pipeline,
+        and wrap the text and additional information as dict according to the inference task type.
+        """
+        output_dicts = []
+        for input_data, output_tensor in zip(data, output):
+            # Get token usage
+            inputs = self.pipeline.tokenizer(
+                input_data,
+                return_tensors=self.pipeline.framework,
+                max_length=self.model_config.get("max_length", None),
+            )
+            prompt_tokens = inputs["input_ids"].shape[-1]
+            total_tokens = len(output_tensor)
+            completions_tokens = total_tokens - prompt_tokens
+
+            # Decode output from tensor
+            generated_text = self.pipeline.tokenizer.decode(
+                output_tensor,
+                skip_special_tokens=True,
+                clean_up_tokenization_spaces=True,
+            )
+
+            # Strip prompt from output
+            completions_text = generated_text[len(input_data) :].lstrip()
+
+            # Determine finish reason
+            finish_reason = "stop"
+            if total_tokens > self.model_config.get(
+                "max_length", float("inf")
+            ) or completions_tokens == self.model_config.get("max_new_tokens", float("inf")):
+                finish_reason = "length"
+
+            # Construct output dict
+            output_dict = {
+                "finish_reason": finish_reason,
+                "usage": {
+                    "prompt_tokens": prompt_tokens,
+                    "completion_tokens": len(output_tensor) - prompt_tokens,
+                    "total_tokens": len(output_tensor),
+                },
+            }
```

Comment:
Can you make these steps into single method like `get_token_usage`, `determine_finish_reason` (probably worth moving to separate file as `__init__.py` becomes way too big already...) and add unit tests for each method + entire postprocess?  This is complex logic so error prone when updating in the future.